### PR TITLE
Index tables

### PR DESCRIPTION
**Pull request by stefpiatek**
Add slow indexes, might be a bit overkill on the location visits but this was from the slow queries.

### CODE REVIEW

File: inform-db/src/main/java/uk/ac/ucl/rits/inform/informdb/movement/LocationVisit.java [hunk 1]
```diff
@@ -33,7 +34,8 @@
  */
 @SuppressWarnings("serial")
 @Entity
-@Table
+@Table(indexes = {@Index(name = "visit_discharge_time", columnList = "hospitalVisitId, dischargeTime DESC"),
+        @Index(name = "location_visit_discharge_time", columnList = "locationId, hospitalVisitId, dischargeTime DESC")})
```
Comment (klapaukh) [inline inform-db/src/main/java/uk/ac/ucl/rits/inform/informdb/movement/LocationVisit.java:38]: So these indexes seem like overkill maybe? Maybe we just need a "is null discharge time" index or something for now? That's just a guess. These may be exactly what is needed. What is their use case?

File: inform-db/src/main/java/uk/ac/ucl/rits/inform/informdb/movement/LocationVisit.java [hunk 2]
```diff
@@ -51,7 +53,7 @@
 
     private Long              parentLocationVisitId;
 
-    @Column(columnDefinition = "timestamp with time zone")
+    @Column(columnDefinition = "timestamp with time zone", nullable = false)
```
Comment (klapaukh) [inline inform-db/src/main/java/uk/ac/ucl/rits/inform/informdb/movement/LocationVisit.java:56]: If you create a location visit from a mid stream A08 it won't have an admission time will it?

**Review by klapaukh — CHANGES_REQUESTED** [review]

File: inform-db/src/main/java/uk/ac/ucl/rits/inform/informdb/movement/LocationVisit.java (hunk 1 above)
Comment (stefpiatek) [reply inform-db/src/main/java/uk/ac/ucl/rits/inform/informdb/movement/LocationVisit.java:38]: yeah that's fair. there are two selects that use either `hospitalVisitId = ? AND dischargeDateTime IS NULL` and `locationVisitId = ?, hospitalVisitId = ? AND dischargeDateTime IS NULL` so yeah I think a null discharge time. I assume that null makes it particularily slow in a query, or is it work indexing the hospitalVisitId as well?

File: inform-db/src/main/java/uk/ac/ucl/rits/inform/informdb/movement/LocationVisit.java (hunk 2 above)
Comment (stefpiatek) [reply inform-db/src/main/java/uk/ac/ucl/rits/inform/informdb/movement/LocationVisit.java:56]: Ah yes, that's a good point, I'll revert that change

File: inform-db/src/main/java/uk/ac/ucl/rits/inform/informdb/movement/LocationVisit.java (hunk 1 above)
Comment (stefpiatek) [reply inform-db/src/main/java/uk/ac/ucl/rits/inform/informdb/movement/LocationVisit.java:38]: Is there a clean way to add a partial index for is null in hibernate, or is this a point where we should be using liquibase?

Comment (klapaukh) [reply inform-db/src/main/java/uk/ac/ucl/rits/inform/informdb/movement/LocationVisit.java:38]: So you don't need to have an index over all the conditions. It's ok to still have a linear search bit as long as that part is small. Ideally we shouldn't have any composite keys unless single column ones don't work I think.

Comment (stefpiatek) [reply inform-db/src/main/java/uk/ac/ucl/rits/inform/informdb/movement/LocationVisit.java:38]: okay, so I'll creata an index for the hospitalVisit and locationId and see where that gets us

**Review by klapaukh — APPROVED** [review]

**Pull request merged**